src/main/java/org/jahia/modules/sitemap/filter/SitemapCacheFilter.java

```diff
@@ -0,0 +1,166 @@
+import org.jahia.services.cache.CacheHelper;
+import org.jahia.services.content.JCRCallback;
+import org.jahia.services.content.JCRNodeWrapper;
+import org.jahia.services.content.JCRSessionWrapper;
+import org.jahia.services.content.JCRTemplate;
+import org.jahia.services.render.RenderContext;
+import org.jahia.services.render.Resource;
+import org.jahia.services.render.filter.AbstractFilter;
+import org.jahia.services.render.filter.RenderChain;
+import org.jahia.services.render.filter.RenderFilter;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.jcr.RepositoryException;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+
+/**
+ * Filter that creates sitemap file nodes for caching.
+ *
+ * Since it can be resource-intensive to go through and generate all sitemap entries,
+ * and since the built-in Jahia view caching isn't guaranteed (i.e. depending on traffic, less-used caches can get invalidated at any time)
+ * we've added a custom file-based caching layer to the sitemap.xml views that will be invalidated only after expiration has passed
+ * (currently set at 4 hours).
+ */
+@Component(service = RenderFilter.class)
+public class SitemapCacheFilter extends AbstractFilter {
+
+    private static final Logger logger = LoggerFactory.getLogger(SitemapCacheFilter.class);
+
+    private static final long EXPIRATION = ConfigServiceUtils.getCacheDuration();
+
+    /** Session attribute flag to check in the sitemap views whether view needs to re-render or not */
+    private static final String RENDER_FLAG_ATTR = "refreshSitemapSession";
+
+
+    @Activate
+    public void activate() {
+        setPriority(15f);
+        setApplyOnNodeTypes("jseont:sitemapResource,jseont:sitemap");
+        setApplyOnModes("live");
+        setDescription("Filter for creating sitemap file nodes for caching");
+        logger.debug("Activated FileCacheFilter");
+    }
+
+    /**
+     * Check if view needs to be re-rendered or not and add a session flag that the view can then check to determine that logic.
+     * Also flush cache here if view needs to be re-rendered.
+     */
+    @Override public String prepare(RenderContext renderContext, Resource resource, RenderChain chain) throws Exception {
+        if (!needsCaching(resource)) return null;
+
+        JCRNodeWrapper cacheNode = getCacheNode(resource.getNode());
+        boolean refreshSitemap = (cacheNode == null) || CacheUtils.isExpired(cacheNode, EXPIRATION);
+        renderContext.getRequest().getSession().setAttribute(RENDER_FLAG_ATTR, refreshSitemap);
+        if (refreshSitemap) { // manually flush jahia cache prior to render
+            CacheHelper.flushOutputCachesForPath(resource.getPath(), false);
+        }
+        return null;
```

Comment:
This one should at some point return something else than null, if we want any cache to be enabled.  Returning null will let the normal rendering of the jsp.

Reply:
Changed logic to return non-null string (empty string) if we are suppose to use cache.

I was using a session attribute then checking on the JSP to make sure we're not re-rendering. Wasn't aware we can also use return value to do the same thing, thanks.